<a href="https://colab.research.google.com/github/SofiaT08/Business-Intelligence-/blob/main/Reto_1_NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Integrantes:**
- Maria Jose Botero Zapata
- Manuela Velez Sosa
- Sofia Torres Roldan

0. Cargar las liberias de trabajo

In [ ]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix

from google.colab import drive
drive.mount('/content/drive')

1. Cargar las bases de datos

In [ ]:
nxl= '/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl)
print(XDB)

#Se seleccionan las variables de interés
XDB=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income','Urban']]
XDB.head(10)

#Se separan las variables de entrada y salida
XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB[['Urban']] #Variable de Salida - Variable de Pronóstico

num_datos = len(XD)
print(num_datos)

          ID  Age  Urban  Mother_Edu  Father_Edu  Siblings  Black  Hispanic  \
0          1   21    1.0         8.0         8.0       1.0    0.0       0.0   
1          2   20    1.0         5.0         8.0       8.0    0.0       0.0   
2          3   18    1.0        10.0        12.0       3.0    0.0       0.0   
3          4   17    1.0        11.0        12.0       3.0    0.0       0.0   
4          5   20    1.0        12.0        12.0       1.0    0.0       0.0   
...      ...  ...    ...         ...         ...       ...    ...       ...   
12681  12682   20    0.0        12.0        12.0       8.0    0.0       0.0   
12682  12683   20    1.0        12.0        14.0      14.0    0.0       0.0   
12683  12684   19    1.0         9.0         8.0       3.0    0.0       0.0   
12684  12685   22    1.0         8.0         NaN       7.0    1.0       0.0   
12685  12686   19    1.0         9.0         NaN       4.0    0.0       0.0   

       White  Christian  ...  Height  Weight  Outgo

2. Eliminar los datos faltantes

In [ ]:
XD.isnull().sum()  # Para ver la cantidad de datos faltantes en cada columna
yd.isnull().sum()
XD = XD.dropna(subset=['Age', 'Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income'])
yd = yd.dropna(subset=['Urban'])
   #Aseguramos la coincidencia entre XD y yd
XD, yd = XD[XD.index.isin(yd.index)], yd[yd.index.isin(XD.index)]

print(XD)
print(yd)


       Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  
1          0.0  
5      40000.0  


3. Número de datos

In [ ]:
num_datos = len(XD)
print(num_datos)

5948


5. ¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

In [ ]:
category_counts = XDB['Urban'].value_counts()
category_percentages = (category_counts / len(XDB)) * 100
print(category_percentages)

Urban
1.0    78.590572
0.0    21.094120
Name: count, dtype: float64


6. Llamamos el modelo

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB ()
mnb.fit(XD,yd)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GaussianNB()

7. Media, Varianza, Desviación y límites de las ocho variables



In [ ]:
#Obtenemos la información del modelo
u=mnb.theta_ #Obtenemos la media
print("la media la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : \n", u)

var=mnb.var_
print("La varianza de la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : \n", var)
sigma=np.sqrt(var)
print("La desviación estándar de la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : \n", sigma)

#Se determinan los límites superiores e inferiores de las tres variables
LS=u+sigma
LI=u-sigma

print("El limite superior de las ocho variables es \n", LS)
print("El limite inferior de las ocho variables es \n", LI)

la media la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : 
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La varianza de la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : 
 [[6.11377773e+00 9.96419474e+00 1.51194652e+01 6.51406931e+00
  1.77789403e+01 1.66650620e+01 9.21039647e+02 1.00161093e+09]
 [6.11792319e+00 1.15645045e+01 1.69373304e+01 5.96364977e+00
  1.76201044e+01 1.76562243e+01 9.01712618e+02 1.30246862e+09]]
La desviación estándar de la Edad,Educación de la Madre, Educación del Padre,Tamaño de la Familia,Autoestima,Altura,Peso, Ingresos es : 
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.1

10. Evaluamos el desempeño del modelo

In [ ]:
from sklearn.metrics import confusion_matrix


ydp2=mnb.predict(XD)
cm=confusion_matrix(yd,ydp2)
print("La matriz de confusión es:\n",cm)
VN=cm[0,0] ; FP=cm[0,1]; FN=cm[1,0] ;VP=cm[1,1]

#Metricas de Desempeño
Ex=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud del modelo es:\n",Ex)
TE = (FP + FN) / (VP + VN + FP + FN)
print("La tasa de error del modelo es:\n",TE)

Se= VP/(VP+FN)
print("La sensibilidad del modelo es:\n",Se)

Sp=VN/(VN+FP)
print("La especificidad del modelo es:\n",Sp)

Pr= VP/(VP+FP)
print("La precisión del modelo es:\n",Pr)


La matriz de confusión es:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
La tasa de error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367


11. Indicar si vive en la ciudad o las áreas rurales.

In [ ]:
XDS=[[34,12,15,2,24,69,160,48000]]
ydp= mnb.predict(XDS) #Se obtiene la predicción
print("La predicción del solicitante es:\n",ydp)

if ydp==0:
  print("La persona en consideración vive en las áreas rurales")

if ydp==1:
  print("La persona en consideración vive en la ciudad")

La predicción del solicitante es:
 [1.]
La persona en consideración vive en la ciudad


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**Análisis de datos:**

La base de datos suministrada contenía un total de 12.686 registros. Sin embargo, al eliminar los datos vacíos, la cantidad se redujo a 5.948, lo que indica que aproximadamente el 47% de los datos fueron excluidos. Esto sugiere que la recolección de la información pudo haber sido ineficiente.
Además, los datos estaban categorizados en Vivienda Urbana (78%) y Vivienda Rural (22%). Para garantizar un óptimo funcionamiento del modelo Naive Bayes, fue necesario trabajar únicamente con los datos completos.


Encontramos que, en promedio, las personas tienen una edad de 18 años. La varianza fue de 4.88, lo que indica que no hay una gran dispersión en las edades respecto al promedio. La desviación estándar es de 2.21 años, lo que significa que la mayoría de las edades se encuentran aproximadamente entre 2.21 años por encima o por debajo de la media. Las personas incluidas en esta recolección de datos son bastante jóvenes, ya que el límite inferior es de
 16.15 años y el límite superior es de 20.57 años.

Es posible evidenciar que los ingresos de la población encuestada tienen una media de 45000 y su desviación estándar es de 10000. Lo anterior, nos da a entender que el rango de ingresos es a partir de 35.000 hasta 55.000.
Al momento de analizar el tamaño de familia, se evidencia que en promedio las familias son de 3 miembros, encontrando una desviación estándar de 1.41 personas. Dando a entender que hay un rango aproximado entre 2 a 4 personas por familia.

Al analizar la variable de Autoestima, se encuentra la media de autoestima es de 20, con una desviación estándar de 4.90. Esto significa que más de la mayoría de personas poseen un rango 15 y 25.
La media de la educación de la madre es de 10 años con una desviación estándar de 3.16 años. La mayoría de las madres tienen entre 7 y 13 años de educación.
En promedio de la educación del padre es de 12 años, con una desviación estándar de 3.54 años. Esto indica que la mayoría de los padres tienen entre 9 y 15 años de educación.

En promedio, las personas tienen un peso de 145.64 lb. La varianza fue de 16.23, y la desviación estándar es de 4.03 lb, lo que significa que la mayoría de los pesos se encuentran aproximadamente 4.03 lb por encima o por debajo de la media. El grupo es relativamente homogéneo en términos de peso, ya que el límite inferior es de 141.61 lb y el límite superior es de 149.67 lb.

En cuanto a la altura, el promedio es de 18.36 pies. Hay una variabilidad moderada en las alturas, con una varianza de 16.23 y una desviación estándar de 4.03. El límite inferior es de 62.95 pies y el límite superior, de 71 pies.
De manera general, el modelo alcanzó una exactitud del 79%, lo que indica un nivel de confiabilidad aceptable para su aplicación en la clasificación de datos. Este resultado sugiere que el modelo tiene un buen desempeño en la predicción de viviendas rurales o urbanas.

Por otro lado, la especificidad del modelo es notablemente alta, alcanzando un 95%. Esto significa que el modelo tiene una gran capacidad para identificar correctamente a las personas que residen en zonas rurales. Gracias a esto, se logra una clasificación más precisa, permitiendo que el modelo diferencie con mayor facilidad entre quienes viven en zonas urbanas y rurales.

Las diferencias que encontramos entre las personas y el tipo de zona en la que habitan son las siguientes:

En primer lugar, en las zonas urbanas, en promedio, se observa un mayor nivel socioeconómico, lo que se refleja en ingresos más altos y, en consecuencia, en un costo de vida más elevado. Además, los padres suelen contar con un mayor capital social, ya que, en general, poseen un nivel educativo más alto.

Por otro lado, en las zonas rurales, las familias tienden a ser más numerosas, posiblemente influenciadas por la naturaleza de los trabajos que allí se desempeñan.

En cuanto a las características físicas, los datos no muestran una variación significativa, por lo que no se identifican diferencias relevantes en este aspecto.
